packages/payment-experience-api/src/api/createRefundController.ts

```diff
@@ -0,0 +1,175 @@
+  getPaidPaymentAdmin,
+  paidPaymentExists,
+  RefundGateway,
+} from '@verkkokauppa/payment-backend'
+import { calculateTotalsFromItems } from '../../../order-experience-api/dist/lib/totals'
+
+const requestSchema = yup.object().shape({
+  params: yup.object().shape({
+    orderId: yup.string().required(),
+  }),
+  headers: yup.object().shape({
+    'api-key': yup.string().required(),
+    namespace: yup.string().required(),
+  }),
+})
+
+export class CreateRefundController extends AbstractController<
+  typeof requestSchema
+> {
+  protected readonly requestSchema = requestSchema
+
+  constructor(
+    private readonly opts: {
+      confirmAndCreatePayment: boolean
+    }
+  ) {
+    super()
+  }
+
+  protected async implementation(
+    req: ValidatedRequest<typeof requestSchema>,
+    res: Response
+  ): Promise<any> {
+    const {
+      params: { orderId },
+      headers: { 'api-key': apiKey, namespace },
+    } = req
+
+    await validateApiKey({ namespace, apiKey })
+
+    const refunds: (Refund & {
+      items: RefundItem[]
+      confirmationUrl?: string
+      payment?: unknown
+    })[] = []
+
+    const errors: ExperienceError[] = []
+
+    const order = await getOrderAdmin({ orderId })
+    const paidPaymentExistsForOrder = await paidPaymentExists({
+      orderId: order.orderId,
+      namespace: order.namespace,
+      user: order.user,
+    })
+
+    const items = order.items
+    try {
+      if (!paidPaymentExistsForOrder) {
+        throw new RequestValidationError(`order ${orderId} must be paid first`)
+      }
+
+      const orderRefunds = await getRefundsByOrderAdmin({ orderId })
+
+      const orderItems = items.map((item: OrderItem) => {
+        const refundedItems = orderRefunds
+          .reduce((a, r) => a.concat(r.items), [] as RefundItem[])
+          .filter((refundItem) => refundItem.orderItemId === item.orderItemId)
+
+        const refundedQuantity = refundedItems.reduce(
+          (quantity, refundItem) => quantity + refundItem.quantity,
+          0
+        )
+
+        if (item.quantity + refundedQuantity > item.quantity) {
+          throw new RequestValidationError(
+            `refunded quantity (now: ${item.quantity}, previously: ${refundedQuantity}) cannot exceed orderItem ${item.orderItemId} quantity ${item.quantity}`
+          )
+        }
+
+        const quantity = item.quantity
+        const { priceNet, priceVat, priceGross } = item
+
+        return {
+          ...item,
+          rowPriceNet: (parseFloat(priceNet) * quantity).toString(),
+          rowPriceVat: (parseFloat(priceVat) * quantity).toString(),
+          rowPriceTotal: (parseFloat(priceGross) * quantity).toString(),
+        }
+      })
+
+      const { refund, items: refundItems } = await createRefund({
+        order: {
+          ...order,
+          ...calculateTotalsFromItems({ items: orderItems }),
+          items: orderItems,
+        },
+      })
+
+      if (!this.opts.confirmAndCreatePayment) {
```

Comment:
Should this variable to be renamed?

Reply:
And is this even required here? Just remove if statement and delete?

Reply:
After a taking a closer look, the variable is always set to true in the routes.ts so this check has no real value being here - most likely a leftover from order-exp createRefund. I'll remove this check and the parameter altogether.